pkg/reconciler/finalizer/finalizer.go

```diff
@@ -0,0 +1,605 @@
+			// Finalize in a background thread, otherwise the watch with time out and be interrupted.
+			go eh.finalize(ctx, obj, metadata.DeletionPropagationPolicyForeground)
+		}
+	} else {
+		if err := eh.reconcileFinalizer(ctx, obj); err != nil {
+			klog.Error("Failed to reconcile finalizer: %v", err)
+			return
+		}
+	}
+}
+
+// OnUpdate adds/removes the finalizer as-needed, depending on the annotation,
+// and handles triggering the finalizer if the deletionTimestamp and finalizer
+// are set.
+func (eh *eventHandler) OnUpdate(oldIObj, newIObj interface{}) {
+	newObj, ok := newIObj.(*unstructured.Unstructured)
+	if !ok {
+		panic(fmt.Sprintf("OnUpdate received unexpected object type %T", newIObj))
+	}
+	klog.V(1).Infof("OnUpdate %s", log.AsJSON(newObj))
+
+	obj := newObj.DeepCopy()
+	ctx := eh.context
+
+	if !obj.GetDeletionTimestamp().IsZero() {
+		// Object being deleted.
+		if controllerutil.ContainsFinalizer(obj, metadata.ReconcilerFinalizer) {
+			// Finalize in a background thread, otherwise the watch with time out and be interrupted.
+			go eh.finalize(ctx, obj, metadata.DeletionPropagationPolicyForeground)
+		}
+	} else {
+		if err := eh.reconcileFinalizer(ctx, obj); err != nil {
+			klog.Error("Failed to reconcile finalizer: %v", err)
+			return
+		}
+	}
+}
+
+// OnDelete handles the case when the object is newly NotFound, after finalizers
+// have run.
+func (eh *eventHandler) OnDelete(oldIObj interface{}) {
+	oldObj, ok := oldIObj.(*unstructured.Unstructured)
+	if !ok {
+		panic(fmt.Sprintf("OnDelete received unexpected object type %T", oldIObj))
+	}
+	klog.V(1).Infof("OnDelete %s", log.AsJSON(oldObj))
+
+	// Do nothing. After the RSync is NotFound is too late to finalize.
+	// The finalizer should have run already in OnUpdate when the deletionTimestamp was added.
+	// We _could_ run the destroyer again to be sure, but the reconciler-manager
+	// will probably stop the reconciler before it could finish. So don't bother.
+}
+
+func (eh *eventHandler) finalize(ctx context.Context, obj *unstructured.Unstructured, policy metadata.DeletionPropagationPolicy) {
+	if policy != metadata.DeletionPropagationPolicyForeground {
+		// Orphan (default policy) needs no cleanup
+		return
+	}
+
+	// Dedupe this function, in case it's triggered while running.
+	if !atomic.CompareAndSwapUint32(&eh.finalizing, 0, 1) {
+		// finalizing != 0
+		// Another finalize thread is already running.
+		// We don't need more than one at a time.
+		return
+	}
+	defer func() {
+		// Reset counter to indicate the background thread is complete.
+		// If it failed, another attempt will be made after the next update.
+		// TODO: do we just want to retry forever instead? If so, we don't need to reset this, right?
+		atomic.CompareAndSwapUint32(&eh.finalizing, 1, 0)
+	}()
+
+	// Stop the parser & remediator
+	klog.V(1).Info("Finalizer scheduled: Parser & Remediator stopping")
+	eh.cancel()
+
+	// TODO: dedupe finalize threads
```

Comment:
Ah, this TODO is stale. I already added deduping to the `finalize` function. The goal was to make sure we didn't have multiple destroyers running simultaneously. But instead of just using locks to delay start of the next one, I used a CompareAndSwap directly (which is what mutex uses under the hood) to discard new `finalize` calls that trigger when `finalize` is already running.

Reply:
I removed the dedupe code. The controller-manager now handles deduping with `MaxConcurrentReconciles: 1` and the `ResourceVersionChangedPredicate`.